File: Monotonic_Stack_29_Oct_2023/0496-next-greater-element-i.go
package main

func nextGreaterElement(nums1 []int, nums2 []int) []int {
    next := make(map[int]int)
    s := make([]int, 0, len(nums2))

    for _, v := range nums2 {
        next[v] = -1

        for top := len(s)-1; top >= 0 && s[top] < v; top-- {
            next[s[top]] = v
            s = s[:top]
        }

        s = append(s, v)
    }

    for i, v := range nums1 {
        nums1[i] = next[v]
    }

    return nums1
}

File: Monotonic_Stack_29_Oct_2023/0496_next_greater_element_i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nextGreater = {}
        s = []
        
        for i, v in enumerate(nums2):
            nextGreater[v] = -1
            while s and s[-1] < v:
                k = s.pop()
                nextGreater[k] = v
            s.append(v)

        result = []
        for k in nums1:
            result.append(nextGreater[k])

        return result
